### PR TITLE
[extractor/youtube] Add support for extracting shorts audio pivot feed

### LINKED ISSUE #4911: [YouTube] YouTube audios
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm requesting a site-specific feature
- [X] I've verified that I'm running yt-dlp version **2022.09.01** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [ ] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

California, USA

### Example URLs

https://youtu.be/2GtVksBMYFM
https://youtu.be/Lyj-MZSAA9o
https://youtube.com/shorts/OgEdGWmlRh8?feature=share

### Provide a description that is worded well enough to be understood

YouTube on mobile provides a feature to make shorts with “sounds” and view existing shorts with sounds on them. Please add support for downloading a list of shorts using a videos sounds. As YouTube does not provide a way to view sounds in HTML5, even on mobile HTML5 and grab their URL, we need to put it like ytsound:https://youtu.be/2GtVksBMYFM

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[Documents]$ yt-dlp -vU ytsound:https://youtu.b
e/2GtVksBMYFMG -s 
[debug] Command-line config: ['-vU', 'ytsound:h
ttps://youtu.be/2GtVksBMYFMG', '-s']
[debug] Encodings: locale UTF-8, fs utf-8, pref
 UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.09.01 [5d7c7d6] (pi
p)
[debug] Python 3.9.10+ (CPython 64bit) - macOS-
15.4.1-iPhone14,5-arm-64bit 
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg n5.0.1-4-ga5ebb3d2
5e (setts), ffprobe n5.0.1-4-ga5ebb3d25e, phant
omjs present, rtmpdump present
[debug] Optional libraries: Crypto-3.15.0, cert
ifi-2022.06.15, mutagen-1.45.1, sqlite3-2.6.0, 
websockets-10.3
[debug] Proxy map: {}
[debug] Loaded 1670 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.09.01, Current version: 2022.09.01
yt-dlp is up to date (2022.09.01)
[debug] [generic] Extracting URL: ytsound:https://youtu.be/2GtVksBMYFMG
[generic] 2GtVksBMYFMG: Downloading webpage
ERROR: [generic] Unable to download webpage: <urlopen error unknown url type: ytsound> (caused by URLError('unknown url type: ytsound'))
  File "/var/mobile/Containers/Data/Application/87BE3170-A2E1-4D77-98F2-3C7FE2FA6E42/Library/lib/python3.9/site-packages/yt_dlp/extractor/common.py", line 670, in extract
    ie_result = self._real_extract(url)
  File "/var/mobile/Containers/Data/Application/87BE3170-A2E1-4D77-98F2-3C7FE2FA6E42/Library/lib/python3.9/site-packages/yt_dlp/extractor/generic.py", line 2641, in _real_extract
    full_response = self._request_webpage(url, video_id, headers={'Accept-Encoding': '*'})
  File "/var/mobile/Containers/Data/Application/87BE3170-A2E1-4D77-98F2-3C7FE2FA6E42/Library/lib/python3.9/site-packages/yt_dlp/extractor/common.py", line 821, in _request_webpage
    raise ExtractorError(errmsg, cause=err)

  File "/var/mobile/Containers/Data/Application/87BE3170-A2E1-4D77-98F2-3C7FE2FA6E42/Library/lib/python3.9/site-packages/yt_dlp/extractor/common.py", line 803, in _request_webpage
    return self._downloader.urlopen(self._create_request(url_or_request, data, headers, query))
  File "/var/mobile/Containers/Data/Application/87BE3170-A2E1-4D77-98F2-3C7FE2FA6E42/Library/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 3656, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/private/var/containers/Bundle/Application/4A7D25F8-7CF3-49FE-B87D-003CD121D26C/a-Shell-mini.app/Library/lib/python3.9/urllib/request.py", line 517, in open
    response = self._open(req, data)
  File "/private/var/containers/Bundle/Application/4A7D25F8-7CF3-49FE-B87D-003CD121D26C/a-Shell-mini.app/Library/lib/python3.9/urllib/request.py", line 539, in _open
    return self._call_chain(self.handle_open, 'unknown',
  File "/private/var/containers/Bundle/Application/4A7D25F8-7CF3-49FE-B87D-003CD121D26C/a-Shell-mini.app/Library/lib/python3.9/urllib/request.py", line 494, in _call_chain
    result = func(*args)
  File "/private/var/containers/Bundle/Application/4A7D25F8-7CF3-49FE-B87D-003CD121D26C/a-Shell-mini.app/Library/lib/python3.9/urllib/request.py", line 1417, in unknown_open
    raise URLError('unknown url type: %s' % type)
urllib.error.URLError: <urlopen error unknown url type: ytsound>
[
```

### PR DESCRIPTION
**Pull request by coletdjnz**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

ADD DESCRIPTION HERE

Closes #4911 

Still need to settle on an extractor name and prefix.

Also there is an issue where if the given video is not the audio of any short then the feed will include itself. This is likely just a bug on YouTube's end and not much we can do (I mean, you'll never run into this case in the app).

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [X] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

**Comment by coletdjnz** [comment]
I was trying to figure out what "sfv_audio_pivot" might mean.

Maybe something like "Shorts that revolve around this one video's audio" (hence "audio pivot")?

**Comment by coletdjnz** [comment]
> > I was trying to think what "sfv_audio_pivot" might mean.
> > 
> > Maybe something like "Shorts that revolve around this one video's audio" (hence "audio pivot")? Or dependency?
> 
> sfv = short-form video ?

hm yeah, that sounds plausible, nice catch

"sfv" comes up quite a bit in various API requests/responses for shorts ui so would make sense.

"reels" is also used in some cases too.

File: yt_dlp/extractor/youtube.py [hunk 1]
```diff
@@ -6307,6 +6317,33 @@ def _real_extract(self, url):
             ie=YoutubeTabIE, video_id=playlist_id)
 
 
+class YoutubeShortsAudioPivotIE(InfoExtractor):
+    IE_DESC = 'YouTube Shorts audio pivot; "ytshortsap:" prefix'
+    IE_NAME = 'youtube:shorts:audiopivot'
+    _VALID_URL = f'(?x)^ytshortsap:{YoutubeIE._VALID_URL[5:]}'  # TODO: might be better to just restrict this to id and /shorts/
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:6323]: I would have restricted it to just the id, but if you want to support URL, might as well support all patterns

File: yt_dlp/extractor/youtube.py [hunk 2]
```diff
@@ -6307,6 +6317,33 @@ def _real_extract(self, url):
...
+        Generates sfv_audio_pivot browse parameter for this video id
+        """
+        token = b'\xf2\x05+\n)\x12\'\n\x0b%b\x12\x0b%b\x1a\x0b%b' % ((video_id.encode(),) * 3)
+        return base64.b64encode(token).decode()
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        return self.url_result(
+            f'https://www.youtube.com/feed/sfv_audio_pivot?bp={self._generate_audio_pivot_params(video_id)}',
+            ie=YoutubeTabIE.ie_key())
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:6344]:
```suggestion
            ie=YoutubeTabIE)
```

File: yt_dlp/extractor/youtube.py [hunk 3]
```diff
@@ -4327,8 +4327,8 @@ def _playlist_entries(self, video_list_renderer):
             yield self._extract_video(renderer)
 
     def _rich_entries(self, rich_grid_renderer):
-        renderer = try_get(
-            rich_grid_renderer, lambda x: x['content']['videoRenderer'], dict) or {}
+        renderer = traverse_obj(
+            rich_grid_renderer, ('content', ['videoRenderer', 'reelItemRenderer']), get_all=False) or {}
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:4331]:
```suggestion
            rich_grid_renderer, ('content', ('videoRenderer', 'reelItemRenderer')), get_all=False) or {}
```

File: yt_dlp/extractor/youtube.py (hunk 1 above)
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:6323]:
```suggestion
    _VALID_URL = f'(?x)ytshortsap:{YoutubeIE._VALID_URL[5:]}'  # TODO: might be better to just restrict this to id and /shorts/
```

**Review by pukkandan — CHANGES_REQUESTED** [review]
I don't really have any opinion on the prefix. Go with what makes sense to you

File: yt_dlp/extractor/youtube.py (hunk 3 above)
Comment (coletdjnz) [reply yt_dlp/extractor/youtube.py:4331]: doesn't this make it a path from videoRenderer->reelItemRenderer?

Comment (pukkandan) [reply yt_dlp/extractor/youtube.py:4331]: No, tuple and list are treated the same

**Pull request merged**